AdminServer/appscale/admin/__init__.py

```diff
@@ -420,14 +458,144 @@ def post(self, project_id, service_id):
+      version_json, _ = self.zk_client.get(version_node)
+    except NoNodeError:
+      raise CustomHTTPError(HTTPCodes.NOT_FOUND, message='Version not found')
 
+    return json.loads(version_json)
+
+  def version_from_payload(self):
+    """ Constructs version from payload.
+
+    Returns:
+      A dictionary containing version details.
+    """
+    update_mask = self.get_argument('updateMask', None)
+    if update_mask is None:
+      message = 'At least one field must be specified for this operation.'
+      raise CustomHTTPError(HTTPCodes.BAD_REQUEST, message=message)
+
+    desired_fields = update_mask.split(',')
+    supported_fields = {'appscaleExtensions.httpPort',
+                        'appscaleExtensions.httpsPort'}
+    for field in desired_fields:
+      if field not in supported_fields:
+        message = ('This operation is only supported on the following '
+                   'field(s): [{}]'.format(', '.join(supported_fields)))
+        raise CustomHTTPError(HTTPCodes.BAD_REQUEST, message=message)
+
+    try:
+      given_version = json_decode(self.request.body)
+    except ValueError:
+      raise CustomHTTPError(HTTPCodes.BAD_REQUEST,
+                            message='Payload must be valid JSON')
+
+    masked_version = utils.apply_mask_to_version(given_version, desired_fields)
+
+    extensions = masked_version.get('appscaleExtensions', {})
+    http_port = extensions.get('httpPort', None)
+    https_port = extensions.get('httpsPort', None)
+    if http_port is not None and http_port not in constants.ALLOWED_HTTP_PORTS:
+      raise CustomHTTPError(HTTPCodes.BAD_REQUEST, message='Invalid HTTP port')
+
+    if (https_port is not None and
+        https_port not in constants.ALLOWED_HTTPS_PORTS):
+      raise CustomHTTPError(HTTPCodes.BAD_REQUEST,
+                            message='Invalid HTTPS port')
+
+    return masked_version
+
+  def update_version(self, project_id, service_id, version_id, new_fields):
+    """ Updates a version node.
+
+    Args:
+      project_id: A string specifying a project ID.
+      service_id: A string specifying a service ID.
+      version_id: A string specifying a version ID.
+      new_fields: A dictionary containing version details.
+    Returns:
+      A dictionary containing completed version details.
+    """
+    version_node = '/appscale/projects/{}/services/{}/versions/{}'.format(
+      project_id, service_id, version_id)
+
+    try:
+      version_json, _ = self.zk_client.get(version_node)
+    except NoNodeError:
+      raise CustomHTTPError(HTTPCodes.NOT_FOUND, message='Version not found')
+
+    version = json.loads(version_json)
+    new_ports = utils.assign_ports(version, new_fields, self.zk_client)
+    version['appscaleExtensions'].update(new_ports)
+    self.zk_client.set(version_node, json.dumps(version))
+    return version
+
+  @gen.coroutine
+  def relocate_version(self, project_id, service_id, version_id, new_fields):
+    """ Assigns new ports to a version.
```

Comment:
It seems that this method does actually more than just relocation of version. Shouldn't it be reflected in method name and docstring?
Or maybe new_fields argument should be replaced with two arguments: http_port, https_port, to make it clear that only ports are updated.